server/src/main/scala/com/cloudera/livy/server/recovery/FileSystemStateStore.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.server.recovery
+
+import java.io.{FileNotFoundException, IOException}
+import java.net.URI
+import java.util
+
+import scala.reflect.ClassTag
+
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.Options.{CreateOpts, Rename}
+import org.apache.hadoop.fs.permission.{FsAction, FsPermission}
+
+import com.cloudera.livy.{LivyConf, Logging}
+import com.cloudera.livy.Utils.usingResource
+
+class FileSystemStateStore(
+    livyConf: LivyConf,
+    mockFileContext: Option[FileContext])
+  extends StateStore(livyConf) with Logging {
+
+  // Constructor defined for StateStore factory to new this class using reflection.
+  def this(livyConf: LivyConf) {
+    this(livyConf, None)
+  }
+
+  private val fsUri = {
+    val fsPath = livyConf.get(LivyConf.RECOVERY_STATE_STORE_URL)
+    require(!fsPath.isEmpty, s"Please config ${LivyConf.RECOVERY_STATE_STORE_URL.key}.")
+    new URI(fsPath)
+  }
+
+  private val fileContext: FileContext = mockFileContext.getOrElse {
+    FileContext.getFileContext(fsUri)
+  }
+
+  {
+    // Only Livy user should have access to state files.
+    fileContext.setUMask(new FsPermission("077"))
+
+    // Create state store dir if it doesn't exist.
+    val stateStorePath = absPath(".")
+    try {
+      fileContext.mkdir(stateStorePath, FsPermission.getDirDefault(), true)
+    } catch {
+      case _: FileAlreadyExistsException =>
+        assert(fileContext.getFileStatus(stateStorePath).isDirectory(),
```

Comment:
it would be better not to use `assert` in source code, user could disable this in the runtime, we could throw an exception instead.